scylla/tests/ccm_integration/ccm/logged_cmd.rs

```diff
@@ -0,0 +1,318 @@
+use tokio::io::{AsyncBufReadExt, BufReader};
+use tokio::process::Command;
+
+/// A simple abstraction to run commands, and emit logs during its execution.
+/// It should allow to run multiple commands in parallel.
+pub(crate) struct LoggedCmd {
+    run_id: AtomicI32,
+}
+
+/// A set of options for the command to run.
+pub(crate) struct RunOptions {
+    /// Environment variables for the command.
+    env: HashMap<String, String>,
+    /// A flag telling whether the command is allowed to fail.
+    /// If set to true, and command fails, the error is not propagated.
+    allow_failure: bool,
+}
+
+impl RunOptions {
+    /// The default run options. With empty environment and `allow_failure` set to false.
+    pub(crate) fn new() -> Self {
+        RunOptions {
+            env: HashMap::new(),
+            allow_failure: false,
+        }
+    }
+
+    pub(crate) fn with_env(mut self, env: HashMap<String, String>) -> Self {
+        self.env = env;
+        self
+    }
+
+    pub(crate) fn allow_failure(mut self, allow: bool) -> Self {
+        self.allow_failure = allow;
+        self
+    }
+}
+
+impl LoggedCmd {
+    pub(crate) async fn new() -> Self {
+        LoggedCmd {
+            run_id: AtomicI32::new(1),
+        }
+    }
+
+    async fn process_child_result(
+        status: io::Result<ExitStatus>,
+        allow_failure: bool,
+        run_id: i32,
+        stderr: Vec<u8>,
+        command_with_args: String,
+    ) -> Result<ExitStatus, Error> {
+        match status {
+            Ok(status) => {
+                match status.code() {
+                    Some(code) => {
+                        tracing::info!("{:15} -> status = {}", format!("exited[{}]", run_id), code);
+                    }
+                    None => {
+                        tracing::info!("{:15} -> status = unknown", format!("exited[{}]", run_id));
+                    }
+                }
+                if !allow_failure && !status.success() {
+                    let tmp = stderr.deref();
+                    return Err(Error::msg(format!(
+                        "Command `{}` failed: {}, stderr: \n{}",
+                        command_with_args,
+                        status,
+                        std::str::from_utf8(tmp)?
+                    )));
+                }
+                Ok(status)
+            }
+            Err(e) => {
+                tracing::info!(
+                    "{:15} -> failed to wait on child process: = {}",
+                    format!("exited[{}]", run_id),
+                    e
+                );
+                Err(Error::from(e).context(format!("Command `{}` failed", command_with_args,)))
+            }
+        }
+    }
+
+    pub(crate) async fn run_command<A, B, C>(
+        &self,
+        command: C,
+        args: A,
+        opts: RunOptions,
+    ) -> Result<ExitStatus, Error>
+    where
+        A: IntoIterator<Item = B> + Clone,
+        B: AsRef<OsStr> + Clone,
+        C: AsRef<OsStr> + Clone,
```

Comment:
:wrench: Let's use better names for type parameters. `A`, `B`, `C` don't say much. I suggest `Args`, `Arg`, `Cmd`.